include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -450,13 +451,32 @@ namespace internal
     {
       Assert(n_rows == n_cols, ExcNotImplemented());
 
-      auto &&transpose_fill_nm = [](Number *           out,
-                                    const Number *     in,
-                                    const unsigned int n,
-                                    const unsigned int m) {
-        for (unsigned int mm = 0; mm < m; ++mm)
-          for (unsigned int nn = 0; nn < n; ++nn)
-            out[mm + nn * m] = *(in++);
+      std::vector<bool> constrained_dofs;
+
+      const auto transpose_fill_nm = [&constrained_dofs](Number *           out,
+                                                         const Number *     in,
+                                                         const unsigned int n,
+                                                         const unsigned int m) {
+        const bool first_round = constrained_dofs.size() == 0;
+
+        if (first_round)
+          constrained_dofs.assign(n, false);
+
+        for (unsigned int mm = 0, c = 0; mm < m; ++mm)
+          for (unsigned int nn = 0; nn < n; ++nn, ++c)
+            {
+              Assert(first_round || (mm != nn) ||
+                       (constrained_dofs[mm] == (in[c] == 0.0)),
+                     ExcInternalError());
+
+              if (mm == nn && in[c] == 0.0)
+                {
+                  out[mm + nn * m]     = 1.0;
+                  constrained_dofs[mm] = true;
+                }
+              else
+                out[mm + nn * m] = in[c];
+            }
       };
```

Comment:
I do not really understand this function. If I read the code correctly, you assume that if the diagonal of the given matrix (either mass matrix or derivative matrix), then the whole row and column need to be zero, so setting the diagonal to 1 is fine. Is this correct? I find this dangerous, even though not mathematically wrong: For a positive definite matrix, which is what we assume here, each diagonal entry needs to be positive. As a result, we could only get a zero if the whole row and column would indeed be zero.

For doing this, I think the code is ugly. Why not simply looping through the diagonal of the mass matrix first and record the constrained entries outside the lambda function, and then use the lambda function that does the copy and sets the diagonal to 1 in both matrices?